job-server/src/main/scala/spark/jobserver/WebApi.scala

```diff
@@ -335,39 +328,42 @@ class WebApi(system: ActorSystem,
         path(Segment) { appName =>
           val timer = binPost.time()
           entity(as[Array[Byte]]) { binBytes =>
-            contentType {
-              case Some(x) =>
+            extractRequest { request =>
                 respondWithMediaType(MediaTypes.`application/json`) { ctx =>
+                  // In test cases entity.contentType is 'application/json' even though the content-type
+                  // header is set. For tests header has to be explicitly extracted
+                  val x: ContentType = request.headers.map({
+                    case `Content-Type`(ct) => Some(ct)
+                    case _ => None
+                  }).find(_.isDefined)
+                    .flatten
+                    .getOrElse(request.entity.contentType)
                   BinaryType.fromMediaType(x.mediaType) match {
                     case Some(binaryType) =>
                       val future = binaryManager ? StoreBinary(appName, binaryType, binBytes)
 
-                      future.map {
+                      future.flatMap {
```

Comment:
The akka-http [Request.Complete](https://doc.akka.io/japi/akka-http/current/akka/http/scaladsl/server/RequestContext.html#complete-akka.http.scaladsl.marshalling.ToResponseMarshallable-) returns a future instead of synchronously producing a result like [spray did](https://github.com/spray/spray/blob/da11335d03898b36040c1ad358084d371b1795d0/spray-routing/src/main/scala/spray/routing/RequestContext.scala#L224). To keep the remaining code unchanged (I had some trouble to prevent _accidental_ indentation changes to prevent the diff from exploding), I resorted to use flatMaps everywhere.